Fix mcrypt fatal - Removed on PHP 7.2

vaurdan opened the pull request:
`mcrypt` library has been deprecated since PHP 7.1, and removed on PHP 7.2. This PR aims to replace the `mcrypt` implementation with the alternative `openssl` library. It also aims to keep backwards compatibility when `mcrypt` is still available on the system. 

There will be issues decrypting old data, however: if the original data was encrypted with `mcrypt`, it will not be compatible with `openssl`, as the Rijndael AES is not available on `openssl` and it's not compatible with `aes-256-cbc` ([source](https://stackoverflow.com/questions/49997338/mcrypt-rijndael-256-to-openssl-aes-256-ecb-conversion)). If there is a way to make the old encrypted data retro compatible, I'm open to suggestions.

I also replaced the usage of `serialize` with `wp_json_encode` for PHPCS compliance. 

### Testing

I did a quick test on my PHP environment (PHP 7.4.21), and the encrypt/decrypt functions are working as intended:

```php
$simple_data = "this is a simple example";
$complex_data = [ 
    'test' => 100, 
    'test_array' => [ 
        'other' => 'test', 
        123
     ]
];
```

Testing the simple string: 

```php
wp> $data = push_syndicate_encrypt( $simple_data );
=> string(60) "b2REUStDTGJISTY1Q0FwRXBJWWNKTkJ1R3Q5ekNSeUl3VUtOQUN5UGJHcz0="
wp> push_syndicate_decrypt ( $data );
=> string(24) "this is a simple example"
```

Testing with a complex array:

```php
wp> $data = push_syndicate_encrypt( $complex_data );
=> string(120) "amRrU1l0Z1dWakdvZ1ByTng2KzNMYWo1QUNjdm5sREVRQmlHQWdIZ0Y2a0ZYcktLNUFwNW5NcDAzblU5d0c2ZzVCdThGOTFLeVMycFJ5YVNMZ0Q1UWc9PQ=="
wp> push_syndicate_decrypt( $data );
=> object(stdClass)#2883 (2) {
  ["test"]=>
  int(100)
  ["test_array"]=>
  object(stdClass)#2884 (2) {
    ["other"]=>
    string(4) "test"
    ["0"]=>
    int(123)
  }
}
```

### TODO/Questions

 * If there is no openssl implementation (or cipher) available, the `push_syndicate_get_cipher` returns false. We should add a warning on the dashboard disclaiming that the access tokens will not be encrypted.
 * Is `aes-256-cbc` the best cipher for this? I believe this is the closest to the original `MCRYPT_RIJNDAEL_256`.
 * How can we handle the potential backwards compatibility issues? Or is it a non-issue?
 
 
This PR will address and close #80, #144, and #150.

GaryJones reviewed (changes requested):
I think this is going to need unit and / integration tests to demonstrate that the encrypt/decrypt functions work on a range of strings, on a range of PHP versions.

> How can we handle the potential backwards compatibility issues?

The plugin's readme.txt currently says it requires WP 3.4 - and that version of WP supports down to PHP 5.2, so ideally all of those versions would be tested in a GitHub Action workflow for the integration tests.

vaurdan commented:
Thanks, @GaryJones! Actually writing a unit test for this functionality is a great idea, and I just added the first draft with 3 tests and 13 assertions. Let me know your thoughts.

My concern with backward compatibility is also on the plugin update action. When a user running PHP 7.0 updates to the new version of Push Syndication, they will have the data encrypted with the old cipher, which makes it just garbage. We might add some code to the `upgrade()` method on the `WP_Push_Syndication_Server` class, but this migration will only be possible if the server is running PHP 7.1 or older, as it will **need** mcrypt to do the upgrade, as the MCRYPT_RIJNDAEL_256 cipher is not available on OpenSSL. ( my idea was to decrypt the option with mcrypt, encrypt it again with openssl, and save )

GaryJones commented on tests/test-encryption.php:
```diff
@@ -0,0 +1,121 @@
+		$encrypted_simple_different = push_syndicate_encrypt( $this->simple_string . '1' );
+		$encrypted_complex          = push_syndicate_encrypt( $this->complex_array );
+
+		// Because older WP versions might use older phpunit versions, assertIsString() might not be available.
+		if ( method_exists( $this, 'assertIsString' ) ) {
```
By including [wp-test-utils](https://github.com/Yoast/wp-test-utils/) package which includes the PHPUnit-Polyfills (see how [wp-parsely](https://github.com/Parsely/wp-parsely/) plugin adds it), then you can use all of the new asserttions up to and including PHPUnit 9, even if running PHPUnit 7.5.

vaurdan replied:
Thanks for the suggestion! I wasn't aware of the polyfill package, super useful. I replaced all the checks for the missing asserts and it's now extending from `WPTestUtils`'s `TestCase` instead.

GaryJones commented on includes/push-syndicate-encryption.php:
```diff
@@ -1,18 +1,72 @@
+
+	// Backwards compatibility with PHP < 7.1. Use mcrypt instead.
+	if ( version_compare( PHP_VERSION, '7.1', '<' ) ) {
+		// @codingStandardsIgnoreStart
+		$data = serialize( $data );
+		return base64_encode(mcrypt_encrypt(MCRYPT_RIJNDAEL_256, md5(PUSH_SYNDICATE_KEY), $data, MCRYPT_MODE_CBC, md5(md5(PUSH_SYNDICATE_KEY))));
+		// @codingStandardsIgnoreEnd
```
Why ignore all coding standards, when you could just ignore the ones we can't address?

```suggestion
		// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize -- kept for BC.
		$data = serialize( $data );
		// phpcs:ignore PHPCompatibility.FunctionUse.RemovedFunctions.mcrypt_encryptDeprecatedRemoved,PHPCompatibility.Extensions.RemovedExtensions.mcryptDeprecatedRemoved,PHPCompatibility.Constants.RemovedConstants.mcrypt_rijndael_256DeprecatedRemoved,PHPCompatibility.Constants.RemovedConstants.mcrypt_mode_cbcDeprecatedRemoved
		return base64_encode( mcrypt_encrypt( MCRYPT_RIJNDAEL_256, md5( PUSH_SYNDICATE_KEY ), $data, MCRYPT_MODE_CBC, md5( md5( PUSH_SYNDICATE_KEY ) ) ) );
```

vaurdan replied:
Ah, I was just trying to avoid a huge line with the PHPCS ignores. But I changed it to be more specific, as suggested.

GaryJones commented on includes/push-syndicate-encryption.php:
```diff
@@ -1,18 +1,72 @@
+	return false; // @TODO: return another default cipher? return exception?
+}
+
 function push_syndicate_encrypt( $data ) {
```
Rather than bulking out this function with the logic for both mcrypt and openssl, I would suggest creating `push_syndicate_encrypt_mcrypt()` and `push_syndicate_encrypt_openssl()`, and then have `push_syndicate_encrypt()` just contain a single condition that defers to one of those new functions. Those new functions can then be tested more in isolation, depending on the PHP version. Likewise for the decrypt function(s).

Alternatively, you can go with a full [Strategy pattern](https://refactoring.guru/design-patterns/strategy/php/example), where the logic for each is stored in their own class, which is then injected as needed. It's a bigger refactor than what you've got here, but a recognised pattern.

vaurdan replied:
That's a great suggestion, thanks. I have rewritten the encryption code to use something similar to the Strategy pattern but using a static class instead. Since the Syndicate_Encryptor (class responsible to encrypt and decrypt the data, similar to the "strategy" class) does not need to be changed on runtime, the `Syndication_Encryption::get_encryptor` class will instantiate and store the best Encryptor, depending on the PHP version.

GaryJones commented on includes/push-syndicate-encryption.php:
```diff
@@ -1,18 +1,72 @@
+function push_syndicate_get_cipher() {
+	$cipher = 'aes-256-cbc';
+
+	if ( version_compare( PHP_VERSION, '7.1', '<' ) ) {
+		return MCRYPT_RIJNDAEL_256;
+	}
```
Instead of testing against a PHP version, why not test against whether the constant is defined or not?

(Think about how we should test for features in a browser, rather than making assumptions about whether features are available or not from the browser user-agent string.)

vaurdan replied:
The issue of testing if `MCRYPT_RIJNDAEL_256` is defined, is that on PHP 7.1 it will return true, but when using the `mcrypt_` functions, a deprecated error will be thrown (`Function mcrypt_encrypt() is deprecated). 7.1 is the middle ground where mcrypt is available, but we don't want to use it because it's deprecated. AFAIK, the best way to prevent this is to check against the PHP version.

GaryJones commented on tests/test-encryption.php:
```diff
@@ -0,0 +1,112 @@
+		);
+	}
+
+	/**
+	 * Runs after the test, clean-up
+	 */
+	public function tearDown() {
+		// Nothing yet.
+	}
```
No need to define an empty method like this - all that's happening is you're stopping any parent tearDown from happening, which could break things.

vaurdan commented:
Thanks for the feedback @GaryJones! I just went ahead and refactored the code with your suggestions. It's possible that I might have overengineered it a little bit, but I'm open to more suggestions. 

I have added a few more test cases, but I was wondering if I should simplify it with smaller tests but with more test cases. What are your thoughts on that?

GaryJones commented on includes/class-syndication-encryptor.php:
```diff
@@ -0,0 +1,27 @@
+abstract class Syndication_Encryptor {
+
+	/**
+	 * @param $data
```
Needs a summary line.

Needs a type (before `$data`), and a description.

Also, can the `@return mixed` be made more specific?

Likewise for other methods.

GaryJones commented on includes/push-syndicate-encryption.php:
```diff
@@ -1,72 +1,26 @@
-	return false; // @TODO: return another default cipher? return exception?
-}
-
+require_once dirname( __FILE__ ) . '/class-syndication-encryption.php';
```
Move this to the root file.

GaryJones reviewed (changes requested):
I like where it's heading, but I've left a few more comments.

re tests - smaller is better, but more focused is even better - instead of only testing `push_syndicate_encrypt`, test the public methods of each individual class instead to test their functionality and responses - and **then** test `push_syndicate_encrypt()` to test the logic which decides which strategy is inserted in which context.

vaurdan commented on includes/class-syndication-encryption.php:
```diff
@@ -0,0 +1,64 @@
+
+require_once dirname( __FILE__ ) . '/class-syndication-encryptor.php';
+require_once dirname( __FILE__ ) . '/class-syndication-encryptor-openssl.php';
+require_once dirname( __FILE__ ) . '/class-syndication-encryptor-mcrypt.php';
```
Yes, that makes sense. I was trying to keep a consistent approach with the remainder of the codebase, but having all the necessary requires/includes on the root file is indeed cleaner.

vaurdan commented:
@GaryJones  thanks for all the feedback! I did another refactor, and changed the `Syndication_Encryptor` to an interface, and moved the encryptor loading logic to the plugin root file, along with all the requires and includes. 

I also followed your suggestion and tried to make the unit tests as specific as possible, and tested every single public method and encryption/decryption functions separately. I ended up doing an abstract `EncryptorTestCase` that should take care of all the basic encrypt/decrypt tests for Encryptors, and each Encryptor test extends from it. 

Let me know your thoughts :)

GaryJones commented on includes/class-syndication-encryption.php:
```diff
@@ -19,22 +15,30 @@ class Syndication_Encryption {
+	 */
+	public static function set_encryptor( $encryptor ) {
+		if ( $encryptor instanceof Syndication_Encryptor ) {
+			self::$encryptor = $encryptor;
```
Since we're now setting state in this object (which encryption and decryption strategy to use, then it would make sense to move away from using static methods everywhere.

Right now with this PR and static methods, someone could do `Syndication_Encryption::encrypt( 'foo' ); and it will generate an error along the lines of "Can't call encrypt() method on `false`.", because there's an unwritten reliance on the `Syndication_Encryption::encrypt()` method only being called after `set_encryptor()`.

Instead, by using non-static methods, we can make use of the class constructor to ensure an encryptor is set. Roughly:

```php
class Syndication_Encryption {
	private $encryptor;

	public function __construct( Syndication_Encryptor $encryptor ) {
		$this->encryptor = $encryptor;
	}

	public function encrypt( $data ) {
		$this->encryptor->encrypt( $data );
	}
	...
}
```
```php
// In composable root.

if ( PHP_VERSION_ID < 70100 ) {
	$syndication_encryption = new Syndication_Encryption( new Syndication_Encryptor_MCrypt() );
} else {
	$syndication_encryption = new Syndication_Encryption( new Syndication_Encryptor_OpenSSL() );
}

// Then available as:
$syndication_encryption->encrypt( 'foo' );
$syndication_encryption->decrypt( $bar );
```

(Note the use of [PHP_VERSION_ID](https://www.php.net/manual/en/function.phpversion.php) which avoids the need for a `version_compare()` function call.)

GaryJones commented on includes/interface-syndication-encryptor.php:
```diff
@@ -0,0 +1,33 @@
+<?php
```
WPCS will probably complain about this file name now, which is why I prefer PSR-4 (and it's needed for test files for PHPUnit 9.1)

GaryJones commented on includes/class-syndication-encryption.php:
```diff
@@ -10,35 +10,36 @@ class Syndication_Encryption {
+	public function set_encryptor( Syndication_Encryptor $encryptor ) {
+		$this->encryptor = $encryptor;
+		return $encryptor;
 	}
```
This setter should be removed - once the Encryption strategy is defined and populated via the constructor, then it should be immutable. We don't want an instance of this class to start encrypting with one strategy and then be changed to another strategy before decryption happens.

vaurdan replied:
My idea of having a setter for the encryption strategy was for the potential migration that will need to happen to migrate the passwords encrypted with mcrypt to openssl. That way, the `set_encryptor` could be used to change from the mcrypt strategy to the openssl strategy during the upgrade process, without the need to create a new instance of `Syndication_Encription`. But since this is an edge case, I think it makes sense to remove the getter and setter and protect the `$encryptor` attribute.

GaryJones replied:
I think for that edge case of the upgrade, you can instantiate two `Syndication_Encryption` objects, one with each strategy, as needed.

GaryJones commented on includes/class-syndication-encryption.php:
```diff
@@ -48,8 +49,8 @@ public static function set_encryptor( $encryptor ) {
-		$encryptor = self::get_encryptor();
+	public function encrypt( $data ) {
+		$encryptor = $this->get_encryptor();
 		return $encryptor->encrypt( $data );
```
Why call a public method when we can just grab the property directly?

```suggestion
		return $this->encryptor->encrypt( $data );
```

GaryJones commented on includes/class-syndication-encryption.php:
```diff
@@ -61,8 +62,8 @@ public static function encrypt( $data ) {
-		$encryptor = self::get_encryptor();
+	public function decrypt( $data, $associative = true ) {
+		$encryptor = $this->get_encryptor();
 		return $encryptor->decrypt( $data, $associative );
```
```suggestion
	public function decrypt( $data, $associative = true ) {
		return $this->encryptor->decrypt( $data, $associative );
```

GaryJones commented on includes/class-syndication-encryption.php:
```diff
@@ -10,35 +10,36 @@ class Syndication_Encryption {
+	public function get_encryptor() {
+		return $this->encryptor;
 	}
 
```
Is this still needed? Why should some external code care what encryption strategy was used? The whole point of the class is to abstract that detail away.

And with the other changes to access the property directly, this method isn't needed internally in the class either.

```suggestion
```

GaryJones commented on tests/test-encryption.php:
```diff
@@ -30,6 +31,9 @@ public function setUp() {
 			3         => true,
 		);
+
+		global $push_syndication_encryption;
+		$this->syndication_encryption = $push_syndication_encryption;
```
Introducing a global into unit tests, seems like doing the opposite of keeping tests isolated?

We shouldn't care what the instantiation looks like in the plugin root file (when comparing `PHP_VERSION_ID`) - we don't need to use that instance.

Instead, these tests should be instantiating `new Syndication_Encryption( new ... )` as needed in each test as part of the _arrange_ step (arrange, act, assert). That way, we're truly testing the code itself, and not how some other part of the plugin may have configured it.

vaurdan replied:
Makes sense. The idea of having a global dependency on a unit test was bugging me, but does that mean that we shouldn't be testing the Encryptor selection logic, that is happening on the plugin bootstrap?

GaryJones replied:
If you feel that should be tested, then you could encapsulate the condition / conditional into its own factory class, and test that.

vaurdan commented:
@GaryJones take a look at the changes. I changed the `EncryptionTest` to stop using the global. However, I left an inline question on your original comment, if you can take a look.

Thank you!

GaryJones commented on tests/test-encryption.php:
```diff
@@ -0,0 +1,104 @@
+		);
+	}
+
+	/**
+	 * Test new instance of Syndication_Encryption with the MCrypt strategy.
+	 *
+	 * @requires extension mcrypt
+	 */
+	public function test_new_encryption_instance_with_mcrypt() {
+		$syndication_encryption = new \Syndication_Encryption( new \Syndication_Encryptor_MCrypt() );
+
+		// Disable deprecated warnings for PHP 7..
+		$original_error_reporting = error_reporting( error_reporting() & ~E_DEPRECATED ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.runtime_configuration_error_reporting
+
+		// Quick encryption/decryption test.
+		$quick_test_string = 'This is a quick test.';
+		$encrypted         = $syndication_encryption->encrypt( $quick_test_string );
+		$decrypted         = $syndication_encryption->decrypt( $encrypted );
+		self::assertEquals( $encrypted, $syndication_encryption->encrypt( $quick_test_string ), 'assert that encryption results are consistent' );
+		self::assertEquals( $quick_test_string, $decrypted, 'assert that decrypted string is the same as original' );
+
+		// Restore original error reporting.
+		error_reporting( $original_error_reporting ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.runtime_configuration_error_reporting
+	}
```
This test is about testing `Syndication_Encryptor_MCrypt()` - whether it can encrypt or decrypt correctly. In which case, you should be able to unwrap the `new Syndication_Encryption()` from around the instantiation in line 40, and move the test method to `tests/test-encryptor-mcrypt.php` and adjust there as needed.

(Likewise for the `test_new_encryption_instance_with_openssl()` below.)

What's needed, is a way to test the `Syndication_Encryption` methods - and that would mean creating a mock object of the interface that it expects to be passed in, and setting what the methods of the mock would return - this way, we can ensure that `Syndication_Encryption::encrypt( 'foo' )` is correctly calling `encrypt()` on whatever instance of `Syndication_Encyryptor` is passed to it.

Untested, but along the lines of:

```php
function test_encrypt_method_is_called_on_encryptor_object() {
	$fake_encrypted_string = 'I am encrypted!';

	$mock_encryptor = $this->createMock( Syndication_Encryptor::class );
	$mock_encryptor->method( 'encrypt' )->will( $this->returnValue( $fake_encrypted_string ) );

	$syndication_encryption = new Syndication_Encryption( $mock_encryptor );
	
	self::assertSame( $fake_encrypted_string, $syndication_encryption->encrypt( 'foo' ) );
}
```

Then add a similar test method for `decrypt()`.

vaurdan replied:
Thank you! I was having some difficulties visualizing what and how I should test in the EncryptionTest and your explanation and example as right on the spot. I now fully understand what you mean by testing the individual methods.

GaryJones reviewed (changes requested):
I think we're getting there.

GaryJones commented on tests/test-encryptor-openssl.php:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Class EncryptorOpenSSLTest
+ */
+class EncryptorOpenSSLTest extends EncryptorTestCase {
```
There's no test for `encrypt()` and `decrypt()` in this class?

Might need to run the tests with `-v` or `-vv` to see a list of the tests and check whether this class automatically calls the `test_simple_encrypt()` and other parent methods or not.

vaurdan replied:
Using `--testdox` (`-v` or `-vv` is not providing any additional output, for some strange reason) I can see that all the tests are running, for each Encryptor:

```
Syndication\Tests\Encryption
 ✔ Encrypt method is called on encryptor object [48.28 ms]
 ✔ Decrypt method is called on encryptor object [0.27 ms]
 ✔ Encryption [2.45 ms]
 ✔ Decryption [0.21 ms]

Syndication\Tests\EncryptorMCrypt
 ✔ Complex encryption [0.21 ms]
 ✔ Complex decryption [0.16 ms]
 ✔ Cipher [0.10 ms]
 ✔ Simple encryption [0.15 ms]
 ✔ Simple decryption [0.14 ms]

Syndication\Tests\EncryptorOpenSSL
 ✔ Cipher [0.90 ms]
 ✔ Simple encryption [0.16 ms]
 ✔ Simple decryption [0.14 ms]
 ✔ Complex encryption [0.15 ms]
 ✔ Complex decryption [0.18 ms]

Time: 6.91 seconds, Memory: 34.00 MB
```

GaryJones commented on tests/class-encryptor-test-case.php:
```diff
@@ -0,0 +1,97 @@
+	/**
+	 * Test an array decryption.
+	 *
+	 * @param string $encrypted The encrypted string from the previous test.
```
I'm not enthralled with this - it seems to make the decryption totally dependent on another test (i.e. it's no isolated like it should be).

We should be able to use a known encrypted string, independent of any other test, set inside the test method, call the decrypt methods and see if it comes out to a known string.

vaurdan replied:
It's tricky to have a known string, because of how OpenSSL implementation works and how it is exposed in the Encryptor methods. To encrypt a string using OpenSSL's AES256 implementation, we need to have an IV (Initialization Vector) and an encryption key. Both the IV and the encryption key depend on the value of `PUSH_SYNDICATE_KEY`, which can (and should) be different from environment to environment. 

This test, by having a dependency, ensures that an encryption is only successful if, with the returned encrypted output, we can decrypt and retrieve the same data as before. If the two methods (encryption and decryption) don't work together (failing the test) effectively the functionality is broken. So technically they are dependent on each other, so are the tests.

---

On a final note, regarding the IV: ideally, it should be unique per encryption, so if the same password is encrypted for different sites, they would have different encrypted outputs. But this would require storing the IV alongside the encrypted password and the current implementation - slightly outside of the scope of this plugin ([here, for example](https://github.com/Automattic/syndication/blob/b47a461679b6c8da8de93c18180aeedc3d4b2c6e/includes/class-syndication-wp-rest-client.php#L252))  - doesn't support this. It's something we can try to improve on a future PR, alongside the version upgrader. Nonetheless, even using the same IV, this implementation using AES256-CBC is strongly recommended over the previous non-standard MCRYPT_RIJNDAEL_256.

GaryJones replied:
Thanks for the explanation - much appreciated.

Since one test is dependent on another then, can you use the [`@depends` annotation](https://phpunit.readthedocs.io/en/9.5/annotations.html#depends) to be more explicit?

vaurdan commented:
@GaryJones is there anything else missing on this one?
i think the `@depends` annotation was already added [here](https://github.com/Automattic/syndication/blob/vip/deprecated-mcrypto/tests/class-encryptor-test-case.php#L77).

GaryJones commented on includes/class-syndication-encryptor-mcrypt.php:
```diff
@@ -0,0 +1,36 @@
+<?php
+
+/**
+ * Class Syndication_Encryptor_OpenSSL
```
```suggestion
 * Class Syndication_Encryptor_MCrypt
```

GaryJones reviewed (commented):
One minor copypasta, and I'd otherwise love to see namespaces used for the new code, but these don't have to be a blocker.

The only consideration would be what happens if someone has data encrypted with mcrypt under PHP 7.1, then updates to PHP 7.2, and finds they can't decrypt using the OpenSSL?

GaryJones commented:
> There will be issues decrypting old data, however: if the original data was encrypted with mcrypt, it will not be compatible with openssl, as the Rijndael AES is not available on openssl and it's not compatible with aes-256-cbc ([source](https://stackoverflow.com/questions/49997338/mcrypt-rijndael-256-to-openssl-aes-256-ecb-conversion)). If there is a way to make the old encrypted data retro compatible, I'm open to suggestions.

This is still an issue.
